fluent-syntax/src/parser.js

```diff
@@ -126,21 +134,26 @@ export default class FluentParser {
       ps.expectChar(ps.current() ? '\n' : undefined);
     }
 
-    if (ps.currentIs('[')) {
-      return this.getSection(ps, comment);
+    if (ps.isIDStart() && (!comment || comment.type === 'Comment')) {
+      return this.getMessage(ps, comment);
     }
 
-    if (ps.isIDStart()) {
-      return this.getMessage(ps, comment);
+    if (ps.currentIs('[')) {
+      this.skipSection(ps);
+      if (comment) {
+        return new AST.GroupComment(comment.content);
+      }
+      return null;
```

Comment:
I might be missing something, but if you bring back the previous order of `ifs`, so first test for `[` and then test for identifier start, you could get away without this `return null` here and just let the code flow to the next `if`.

You could then remove the `node === null` check in `withSpan` and `parse`.

Reply:
>  and just let the code flow to the next if.

Then it'll return regular `Comment` for old style comment that was attached to a section, instead of GroupComment.

> You could then remove the node === null check in withSpan and parse.

That's crashes if getEntry returns `null` (it does when section doesn't have a comment in 0.4 style).

Reply:
I mean something like this:

```js
  getEntry(ps) {
    let comment;

    if (ps.currentIs('/') || ps.currentIs('#')) {
      comment = this.getComment(ps);

      // The Comment content doesn't include the trailing newline. Consume
      // this newline here to be ready for the next entry.  undefined stands
      // for EOF.
      ps.expectChar(ps.current() ? '\n' : undefined);
    }

    if (ps.currentIs('[')) {
      this.skipSection(ps);
      if (comment) {
        return new AST.GroupComment(comment.content);
      }
    }

    if (ps.isIDStart() && (!comment || comment.type === 'Comment')) {
      return this.getMessage(ps, comment);
    }

    if (comment) {
      return comment;
    }

    throw new ParseError('E0002');
  }
```

Then you never return `null`.

Reply:
I do in case the file is:

```
[[ This is a section ]]
```

which is one of our tests :(

What do you prefer here?

Reply:
Ah, good point, thanks. Let's keep the tests.

Can we at least remove the `node === null` check in `withSpans`? I think we only return `null` from `getEntry` which isn't decorated with `withSpans` because it handles spans in an own custom way.